File: packages/geometry-utils/src/bound-rect.ts
import { PointF64 } from './point';

export default class BoundRect {
    private _memSeg: Float64Array;

    private _position: PointF64;

    private _size: PointF64;

    constructor(x: number = 0, y: number = 0, width: number = 0, height: number = 0) {
        this._memSeg = new Float64Array([x, y, width, height]);
        this._position = new PointF64(this._memSeg, 0);
        this._size = new PointF64(this._memSeg, 2);
    }

    public clone(): BoundRect {
        return new BoundRect(this._position.x, this._position.y, this._size.x, this._size.y);
    }

    public update(position: PointF64, size: PointF64): void {
        this._position.update(position);
        this._size.update(size);
    }

    public get position(): PointF64 {
        return this._position;
    }

    public get size(): PointF64 {
        return this._size;
    }

    public get x(): number {
        return this._position.x;
    }

    public get y(): number {
        return this._position.y;
    }

    public get width(): number {
        return this._size.x;
    }

    public get height(): number {
        return this._size.y;
    }
}

File: packages/wasm-nesting/src/geometry/bound_rect.rs
use crate::geometry::point::Point;
use crate::utils::number::Number;

pub struct BoundRect<T: Number> {
    position: Point<T>,
    size: Point<T>,
}

impl<T: Number> BoundRect<T> {
    pub fn new(x: T, y: T, width: T, height: T) -> Self {
        let position = Point::new(Some(x), Some(y));
        let size = Point::new(Some(width), Some(height));

        Self { position, size }
    }

    pub unsafe fn update(&mut self, position: *const Point<T>, size: *const Point<T>) {
        self.position.update(position);
        self.size.update(size);
    }

    pub fn position(&self) -> *const Point<T> {
        &self.position as *const Point<T>
    }

    pub fn size(&self) -> *const Point<T> {
        &self.size as *const Point<T>
    }

    pub unsafe fn x(&self) -> T {
        self.position.x
    }

    pub unsafe fn y(&self) -> T {
        self.position.y
    }

    pub unsafe fn width(&self) -> T {
        self.size.x
    }

    pub unsafe fn height(&self) -> T {
        self.size.y
    }

    // Якщо треба clone (глибоке копіювання)
    pub unsafe fn clone(&self) -> Self {
        Self::new(self.x(), self.y(), self.width(), self.height())
    }
}
